map-pointer command

alex-courtis opened the pull request:
resolves #685

superceeds #691

`map-pointer mode modifiers button action|command`

View under cursor is always focused when mapped button pressed.

alex-courtis commented on river/Seat.zig:
```diff
@@ -419,7 +419,7 @@ pub fn handleSwitchMapping(
 }
 
-fn runCommand(self: *Self, args: []const [:0]const u8) void {
+pub fn runCommand(self: *Self, args: []const [:0]const u8) void {
```
Exposed for convenience.

alex-courtis commented:
### Tested

```
riverctl map-pointer normal Super BTN_LEFT move-view

riverctl map-pointer normal Super BTN_RIGHT resize-view

riverctl map-pointer normal Super BTN_MIDDLE toggle-float

riverctl map-pointer normal Super+Control BTN_MIDDLE toggle-fullscreen

riverctl map-pointer normal Super+Control BTN_MIDDLE swap next

riverctl unmap-pointer normal Super+Control BTN_MIDDLE

riverctl unmap-pointer normal Super+Control BTN_MIDDLE

riverctl map-pointer normal Super+Control BTN_MIDDLE
error: not enough arguments

riverctl map-pointer normal Super+Control BTN_BLEH toggle-fullscreen
error: unknown button BTN_BLEH

riverctl map-pointer normal Super+Control BTN_MIDDLE
error: not enough arguments

riverctl map-pointer normal Super BTN_RIGHT resize-view aoeu
error: too many arguments
```


Fullscreen view move-view resize-view and commands.


Ensured button not sent to client on click


Multiple outputs actions and commands

alex-courtis commented on river/command/map.zig:
```diff
@@ -140,20 +139,18 @@ pub fn mapPointer(
         PointerMapping.Action.move
     else if (mem.eql(u8, args[4], "resize-view"))
```
Parsing move/resize could be moved to `PointerMapping` `init`.

Leon-Plickat commented on river/Cursor.zig:
```diff
@@ -847,8 +849,6 @@ fn surfaceAtFilter(view: *View, filter_tags: u32) bool {
 pub fn enterMode(self: *Self, mode: enum { move, resize }, view: *View) void {
     log.debug("enter {s} cursor mode", .{@tagName(mode)});
 
-    self.seat.focus(view);
```
Hmm... I can not come up with an example right now, but I have the feeling that this might cause some issues somewhere. Button mappings aren't the only thing that can change the cursor mode.

Right now I would say that it's better to keep this `focus()` call here and move the one in `handlePointerMapping()` into the `.command` branch of the switch.

alex-courtis replied:
Safe. Done.

Leon-Plickat commented on river/PointerMapping.zig:
```diff
@@ -14,13 +14,47 @@
+    command_args: []const [:0]const u8,
+) !Self {
+    const args_len = if (action == Action.command) command_args.len else 0;
+    const owned_args = try util.gpa.alloc([:0]u8, args_len);
```
Even for static length arrays, it's probably a good idea to use `ArrayListUnmanaged`, since it's API is a bit more handy.

Roughly like this:
```zig
var owned_args: std.ArrayListUnmanaged([:0]u8) = .{};
try owned_args.ensureTotalCapacity(util.gpa, command_args.len);
errdefer {
    for (owned_args.items) |arg| util.gpa.free(arg);
    owned.deinit(util.gpa);
}
for (command_args) |arg| {
    const owned = try util.gpa.dupeZ(arg);
    owned_args.appendAssumeCapacity(owned);
}
```

It might even make sense to put these allocations into an arena for easier cleanup, maybe (Just make sure to not grow the array list after the initial setup).

alex-courtis replied:
I'm struggling with the use of an arena. Attempt:

```zig
        ActionType.command => blk: {
            var arena = std.heap.ArenaAllocator.init(util.gpa);
            defer arena.deinit();

            var owned_args = try std.ArrayListUnmanaged([:0]const u8).initCapacity(arena.allocator(), command_args.len);

            for (command_args) |arg| {
                const owned = try util.gpa.dupeZ(u8, arg);
                owned_args.appendAssumeCapacity(owned);
            }

            break :blk Action{ .command = owned_args.toOwnedSlice(util.gpa) };
        },
```

Even though the strings and the slice are allocated using gpa the arena always frees everything.

Leon-Plickat replied:
The arena would be part of the `Action` struct, holding the strings as well as the slice. Then in `Action.deinit()` just call `Arena.deinit()` for cleanup.

alex-courtis replied:
That make a lot of sense, thank you.

I'll get to that next week.

alex-courtis replied:
Added the `ArenaAllocator` to `PointerMapping`.

This is a really clean pattern. It can applied to `Mapping` and friends.

Leon-Plickat commented on river/PointerMapping.zig:
```diff
@@ -14,13 +14,47 @@
+    action: Action,
+    command_args: []const [:0]const u8,
+) !Self {
+    const args_len = if (action == Action.command) command_args.len else 0;
```
I'd prefer a cleaner separation of the command and non-command cases.

alex-courtis replied:
Refactored as switch.

Leon-Plickat commented on river/PointerMapping.zig:
```diff
@@ -14,13 +14,47 @@
 event_code: u32,
 modifiers: wlr.Keyboard.ModifierMask,
 action: Action,
+command_args: []const [:0]const u8,
```
You could turn `action` into a tagged union (using `Action` as tag-type). Then `command_args` could be the `command` case, with the two others being `void`.

alex-courtis replied:
That's a nice construct. Done.

Leon-Plickat reviewed (commented):
Took me some time to get to this, but here is a preliminary review. Just a few minor things.

alex-courtis commented:
> Took me some time to get to this, but here is a preliminary review. Just a few minor things.

No worries at all.

Thanks for the zig tips... I'll get to this one early next week.

alex-courtis commented on river/PointerMapping.zig:
```diff
@@ -14,13 +14,72 @@
+                owned_args.appendAssumeCapacity(owned);
+            }
+
+            const action: Action = Action{ .command = owned_args.toOwnedSlice(util.gpa) };
```
I think I have misunderstood your intent here. Did you mean to change the action/command type to an `ArrayList` or use the slice?

Leon-Plickat replied:
No, I think you got it. This is exactly what I recommended: Using an ArrayList instead of doing the slice allocation and population manually.

alex-courtis replied:
Done.

alex-courtis commented on river/PointerMapping.zig:
```diff
@@ -14,13 +14,72 @@
+
+            const action: Action = Action{ .command = owned_args.toOwnedSlice(util.gpa) };
+
+            owned_args.deinit(util.gpa);
```
I do not think this is correct. It seems that a double free may occur from the `errdefer` block.

alex-courtis commented:
This has been a fantastic learning experience.

Thank you for all your help and patience.

alex-courtis commented:
This works. Repeated tests and dogfooding now.

I could not validate the freeing via valgrind. There's a lot of noise from wl_roots which I can [deal with](https://github.com/alex-courtis/way-displays/blob/3d2b4fab16d4c43eada42723fbe61670d096f259/way-displays.supp). Is there a zig tool?

When, say, commenting out the arena.deinit, I did not see anything referencing PointerMapping.

ifreund reviewed (approved):
Thanks for your contribution and sorry for the long wait. I haven't had as much time for river recently as I would like.

I've squashed your commits and added a few fixes/style tweaks in a separate commit myself rather than bothering you with another round of review and more delay.

Assuming the CI is green I'll merge this in a few minutes :)

alex-courtis commented:
> I've squashed your commits and added a few fixes/style tweaks in a separate commit myself rather than bothering you with another round of review and more delay.

Thank you. Much more efficient and my preferred workflow.
